retriever/retriever.go

```diff
@@ -0,0 +1,97 @@
+package retriever
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/data-preservation-programs/singularity/retriever/deserializer"
+	lassietypes "github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/storage"
+	trustlessutils "github.com/ipld/go-trustless-utils"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"go.uber.org/multierr"
+)
+
+type EndpointFinder interface {
+	FindHTTPEndpoints(ctx context.Context, sps []string) ([]peer.AddrInfo, error)
+}
+
+// a retriever returns a byte stream for a cid at the root of a unixfs tree,
+// from a list of Filecoin providers
+type Retriever struct {
+	lassie         lassietypes.Fetcher
+	endpointFinder EndpointFinder
+}
+
+// deserialize takes an reader of a carFile and writes the deserialized output
+func (r *Retriever) deserialize(ctx context.Context, c cid.Cid, rangeStart int64, rangeEnd int64, carInput io.Reader, carOutput io.Writer) error {
+	cr, err := car.NewBlockReader(carInput)
+	if err != nil {
+		return err
+	}
+	_, err = deserializer.Deserialize(ctx, cr, c, rangeStart, rangeEnd, carOutput)
+	return err
+}
+
+// getContent fetches content through Lassie and writes the CAR file to an output writer
+func (r *Retriever) getContent(ctx context.Context, c cid.Cid, rangeStart int64, rangeEnd int64, sps []string, carOutput io.Writer) error {
+	writable, err := storage.NewWritable(carOutput, []cid.Cid{c}, car.WriteAsCarV1(true))
```

Comment:
the round-trip through a CARv1 seems unfortunate, serializing and deserializing when you have the blocks is maybe a tad inefficient? Perhaps you should make this accept a `storage.WritableStorage`, and the "deserializer" accept a `storage.ReadableStorage` and your pipe is just a custom storage implementation that takes blocks from the write side, queues them up and puts them out on the read side, checking the expected read CID matches the queued CID and erroring if not. You could also apply the backpressure there too, preventing writes until you get a read, or however much buffering you want to do.

Reply:
I did something kind of similar here https://github.com/ipld/go-car/blob/master/cmd/car/extract.go#L380C6-L380C22 but it's got to account for the no-dupes case and it _is_ decoding a CAR. But the point is that it's providing a `ReadableStorage` to the unixfs reifier.

Reply:
ah I see what you mean, skip carv1 streaming and jsut make a block pipe? interesting!

Reply:
storage.Pipe seems like an interesting concept thought I wonder if it belongs in go-ipld-prime :P